File: baekjoon/2493.py
import sys

input = sys.stdin.readline

n = int(input())

lst = list(map(int, input().split()))
answer = [0]

for i in range(1, n):
    tmp = i - 1
    while 1:
        if lst[tmp] < lst[i]:
            if answer[tmp] == 0:
                answer.append(0)
                break
            else:
                tmp = answer[tmp] - 1
        else:
            answer.append(tmp + 1)
            break

print(" ".join(map(str, answer)))

File: baekjoon/2493.cpp
#include <cstdio>
#include <vector>
using namespace std;

int main()
{
    int n;
    scanf("%d", &n);
    vector<int> array(n);
    vector<int> answer(n, 0);

    for (int i = 0; i < n; i++)
    {
        scanf("%d", &array[i]);
    }

    for (int i = 1; i < n; i++)
    {
        int tmp = i - 1;
        while (1)
        {
            if (array[tmp] < array[i])
            {
                if (answer[tmp] == 0)
                {
                    answer[i] = 0;
                    break;
                }
                tmp = answer[tmp] - 1;
            }
            else
            {
                answer[i] = tmp + 1;
                break;
            }
        }
    }

    for (int i = 0; i < n; i++)
    {
        printf("%d ", answer[i]);
    }
    printf("\n");

    return 0;
}
